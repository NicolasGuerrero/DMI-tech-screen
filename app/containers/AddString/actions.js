/*
 *
 * AddString actions
 *
 */

import {
  CHANGE_STRING_INPUT,
  ADD_STRING,
  ADD_STRING_SUCCESS,
  ADD_STRING_ERROR,
} from './constants';

export function changeStringInput(newString) {
  console.log('string from actions', newString);
  return {
    type: CHANGE_STRING_INPUT,
    newString,
  };
}

export function addString() {
  return {
    type: ADD_STRING,
  };
}

export function stringAdded(string) {
  return {
    type: ADD_STRING_SUCCESS,
    string,
  };
}

export function stringAddingError(error) {
  return {
    type: ADD_STRING_ERROR,
    error,
  };
}
