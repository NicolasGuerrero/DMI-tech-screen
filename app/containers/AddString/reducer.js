/*
 *
 * AddString reducer
 *
 */
import produce from 'immer';
import {
  CHANGE_STRING_INPUT,
  ADD_STRING,
  ADD_STRING_SUCCESS,
  ADD_STRING_ERROR,
} from './constants';

export const initialState = {
  adding: false,
  error: false,
  added: false,
  string: '',
};

/* eslint-disable default-case, no-param-reassign */
const addStringReducer = (state = initialState, action) =>
  produce(state, draft => {
    switch (action.type) {
      case CHANGE_STRING_INPUT:
        draft.newString = action.newString;
        break;

      case ADD_STRING:
        draft.adding = true;
        draft.error = false;
        break;

      case ADD_STRING_SUCCESS:
        draft.string = action.string;
        draft.adding = false;
        draft.added = true;
        break;

      case ADD_STRING_ERROR:
        draft.error = action.error;
        draft.adding = false;
        break;
    }
  });

export default addStringReducer;
