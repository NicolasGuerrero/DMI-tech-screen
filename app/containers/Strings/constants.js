/*
 *
 * Strings constants
 *
 */

export const DEFAULT_ACTION = 'app/Strings/DEFAULT_ACTION';
