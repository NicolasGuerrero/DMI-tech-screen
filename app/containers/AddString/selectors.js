import { createSelector } from 'reselect';
import { initialState } from './reducer';

/**
 * Direct selector to the addString state domain
 */

const selectAddString = state => state.addString || initialState;

/**
 * Other specific selectors
 */

/**
 * Default selector used by AddString
 */

const makeSelectAddString = () =>
  createSelector(
    selectAddString,
    substate => substate.newString,
  );

const makeSelectAddStringAdding = () =>
  createSelector(
    selectAddString,
    substate => substate.adding,
  );

const makeSelectAddStringAdded = () =>
  createSelector(
    selectAddString,
    substate => substate.added,
  );

const makeSelectAddStringError = () =>
  createSelector(
    selectAddString,
    substate => substate.error,
  );

export {
  selectAddString,
  makeSelectAddString,
  makeSelectAddStringAdding,
  makeSelectAddStringAdded,
  makeSelectAddStringError,
};
