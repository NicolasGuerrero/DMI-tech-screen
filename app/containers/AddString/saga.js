import { call, put, select, takeLatest } from 'redux-saga/effects';
import { ADD_STRING } from 'containers/AddString/constants';
import { stringAdded, stringAddingError } from 'containers/AddString/actions';
import request from 'utils/request';
import { makeSelectAddString } from './selectors';

// Individual exports for testing
export function* addString() {
  // See example in containers/HomePage/saga.js

  const inputString = yield select(makeSelectAddString());
  const requestURL = `/api/strings`;
  const options = {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
    },
    body: JSON.stringify({ inputString }),
  };
  try {
    const addedString = yield call(request, requestURL, options);
    yield put(stringAdded(addedString));
  } catch (err) {
    console.log('from saga', err);
    yield put(stringAddingError(err));
  }
}

/**
 * Root saga manages watcher lifecycle
 */
export default function* addStringData() {
  // Watches for ADD_STRING actions and calls addString when one comes in.
  // By using `takeLatest` only the result of the latest API call is applied.
  // It returns task descriptor (just like fork) so we can continue execution
  // It will be cancelled automatically on component unmount
  yield takeLatest(ADD_STRING, addString);
}
