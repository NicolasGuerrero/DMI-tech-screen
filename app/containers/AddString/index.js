/**
 *
 * AddString
 *
 */

import React, { memo, useEffect } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';

import { createStructuredSelector } from 'reselect';
import { compose } from 'redux';

import { useInjectSaga } from 'utils/injectSaga';
import { useInjectReducer } from 'utils/injectReducer';
import AddStringNotifications from 'components/AddStringNotifications';
import {
  makeSelectAddString,
  makeSelectAddStringAdding,
  makeSelectAddStringAdded,
  makeSelectAddStringError,
} from './selectors';
import reducer from './reducer';
import saga from './saga';
import Form from './Form';
import Input from './Input';
import { changeStringInput, addString, addStringReset } from './actions';

export function AddString({
  newString,
  adding,
  added,
  error,
  onSubmitForm,
  onChangeString,
  reset,
}) {
  useInjectReducer({ key: 'addString', reducer });
  useInjectSaga({ key: 'addString', saga });

  useEffect(() => {
    reset();
  }, []);

  const notificationProps = { adding, added, error };

  return (
    <div>
      <Form onSubmit={onSubmitForm}>
        <label htmlFor="newString">
          <Input
            id="newString"
            type="text"
            placeholder="Enter a string..."
            value={newString}
            onChange={onChangeString}
          />
          <button type="submit">Submit</button>
        </label>
      </Form>
      <AddStringNotifications {...notificationProps} />
    </div>
  );
}

AddString.propTypes = {
  adding: PropTypes.bool,
  added: PropTypes.bool,
  error: PropTypes.oneOfType([PropTypes.object, PropTypes.bool]),
  onSubmitForm: PropTypes.func,
  newString: PropTypes.string,
  reset: PropTypes.func,
  onChangeString: PropTypes.func,
};

const mapStateToProps = createStructuredSelector({
  newString: makeSelectAddString(),
  adding: makeSelectAddStringAdding(),
  added: makeSelectAddStringAdded(),
  error: makeSelectAddStringError(),
});

export function mapDispatchToProps(dispatch) {
  return {
    onChangeString: evt => dispatch(changeStringInput(evt.target.value)),
    onSubmitForm: evt => {
      if (evt !== undefined && evt.preventDefault) evt.preventDefault();
      dispatch(addString());
      dispatch(changeStringInput(''));
    },
    reset: () => dispatch(addStringReset()),
  };
}

const withConnect = connect(
  mapStateToProps,
  mapDispatchToProps,
);

export default compose(
  withConnect,
  memo,
)(AddString);
