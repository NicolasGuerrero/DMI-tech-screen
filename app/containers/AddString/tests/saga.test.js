/**
 * Test sagas
 */

import { put, takeLatest } from 'redux-saga/effects';
import { stringAdded, stringAddingError } from 'containers/AddString/actions';
import { ADD_STRING } from '../constants';

import addStringData, { addString } from '../saga';

const testString = 'Test String.';

describe('addString Saga', () => {
  let addStringGenerator;

  // We have to test twice, once for a successful load and once for an unsuccessful one
  // so we do all the stuff that happens beforehand automatically in the beforeEach
  beforeEach(() => {
    addStringGenerator = addString();

    const selectDescriptor = addStringGenerator.next().value;
    expect(selectDescriptor).toMatchSnapshot();

    const callDescriptor = addStringGenerator.next(testString).value;
    expect(callDescriptor).toMatchSnapshot();
  });

  it('should dispatch the stringAdded action if it requests the data successfully', () => {
    const response = testString;

    const putDescriptor = addStringGenerator.next(response).value;
    expect(putDescriptor).toEqual(put(stringAdded(response, testString)));
  });

  it('should call the stringAddingError action if the response errors', () => {
    const response = new Error('Some error');
    const putDescriptor = addStringGenerator.throw(response).value;
    expect(putDescriptor).toEqual(put(stringAddingError(response)));
  });
});

describe('addStringData Saga', () => {
  const addStringDataSaga = addStringData();

  it('should start task to watch for LOAD_REPOS action', () => {
    const takeLatestDescriptor = addStringDataSaga.next().value;
    expect(takeLatestDescriptor).toEqual(takeLatest(ADD_STRING, addString));
  });
});
