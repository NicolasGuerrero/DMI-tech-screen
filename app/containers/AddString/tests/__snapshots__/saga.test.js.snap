// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`addString Saga should call the stringAddingError action if the response errors 1`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [],
    "selector": [Function],
  },
  "type": "SELECT",
}
`;

exports[`addString Saga should call the stringAddingError action if the response errors 2`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [
      "/api/strings",
      Object {
        "body": "{\\"inputString\\":\\"Test String.\\"}",
        "headers": Object {
          "Content-Type": "application/json",
        },
        "method": "POST",
      },
    ],
    "context": null,
    "fn": [Function],
  },
  "type": "CALL",
}
`;

exports[`addString Saga should dispatch the stringAdded action if it requests the data successfully 1`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [],
    "selector": [Function],
  },
  "type": "SELECT",
}
`;

exports[`addString Saga should dispatch the stringAdded action if it requests the data successfully 2`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [
      "/api/strings",
      Object {
        "body": "{\\"inputString\\":\\"Test String.\\"}",
        "headers": Object {
          "Content-Type": "application/json",
        },
        "method": "POST",
      },
    ],
    "context": null,
    "fn": [Function],
  },
  "type": "CALL",
}
`;
